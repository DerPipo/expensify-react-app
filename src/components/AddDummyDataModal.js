import React, { useState } from 'react';
import Modal from 'react-modal';
import { connect } from 'react-redux';
import createDummyData from '../store/createDummyData';

const AddDummyDataModal = props => {
  const [modalIsOpen, setModalIsOpen] = useState(true);
  const closeModal = () => {
    setModalIsOpen(false);
  };

  const onCreateDummyData = () => {
    createDummyData(props.dispatch);
  };
  return (
    <Modal
      isOpen={modalIsOpen}
      appElement={document.getElementById('app')}
      onRequestClose={closeModal}
      className="modal"
    >
      <h1 className="modal__title">First launch?</h1>
      <p className="modal__text">
        It seems that this is your first launch or you don't have stored any
        data yet. Do you want to begin with some dummy data to see how the app
        feels like before storing your own data? (They won't get stored in your
        database, just refresh the page to start from scratch.)
      </p>
      <div className="modal__button-container">
        <button className="modal__button button" onClick={onCreateDummyData}>
          Yes, please
        </button>
        <button
          className="modal__button button button--secondary"
          onClick={closeModal}
        >
          No, thanks
        </button>
      </div>
    </Modal>
  );
};

const mapStateToProps = state => ({
  hasNoExpenses: state.expenses.length === 0,
});

export default connect(mapStateToProps)(AddDummyDataModal);
