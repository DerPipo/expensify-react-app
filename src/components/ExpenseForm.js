import React from 'react'
import moment from 'moment'
import { SingleDatePicker } from 'react-dates'
// import 'react-dates/initialize'


class ExpenseForm extends React.Component {
    constructor(props) {
        super(props)

        this.state = {
            description: props.expense ? props.expense.description : '',
            amount: props.expense ? (props.expense.amount / 100).toString() : '',
            note: props.expense ? props.expense.note : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calendarFocus: false,
            error: ''
        }
    }
    onDescriptionChange = (e) => {
        const description = e.target.value
        this.setState(() => ({ description }))
    }
    onAmountChange = (e) => {
        const amount = e.target.value
        if (!amount || amount.match(/^\d+(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount }))
        }
        // const amount = parseInt(e.target.value)
    }
    onNoteChange = (e) => {
        const note = e.target.value
        this.setState(() => ({ note }))
    }
    onDateChange = (date) => {
        if (date) {
            this.setState(() => ({ createdAt: date }))
        }
    }
    onCalendarFocusChange = ({  focused }) => {
        this.setState(() => ({ calendarFocus: focused }))
    }
    onSubmit = (e) => {
        e.preventDefault()
        
        if (!this.state.description || !this.state.amount) {
            // Set error state equal to 'Please provide description and amount.'
            this.setState(() => ({ error: 'Please provide description and amount.' }))
        } else {
            this.setState(() => ({ error: '' }))
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount) * 100,
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            })
        }
    }
    render() {
        return (
            <form className="form" onSubmit={this.onSubmit}>
                {this.state.error && 
                    <p className="form__error">{this.state.error}</p>
                }
                <input 
                    className="text-input"
                    type="text" 
                    placeholder="Description" 
                    autoFocus 
                    value={this.state.description} 
                    onChange={this.onDescriptionChange} 
                />
                <input 
                    className="text-input" 
                    type="text" 
                    placeholder="Amount" 
                    value={this.state.amount} 
                    onChange={this.onAmountChange} 
                />
                <SingleDatePicker
                    date={this.state.createdAt}
                    onDateChange={this.onDateChange}
                    focused={this.state.calendarFocus}
                    onFocusChange={this.onCalendarFocusChange} 
                    // id="single_date_picker_id"
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                />
                <textarea 
                    className="textarea"
                    placeholder="Add a note for your expense (optional)" 
                    value={this.state.note} 
                    onChange={this.onNoteChange}>
                </textarea>
                <div>
                    <button className="button" type="submit">{this.props.expense ? 'Save Expense' : 'Add Expense'}</button>
                </div>
            </form>
        )
    }
}

export default ExpenseForm