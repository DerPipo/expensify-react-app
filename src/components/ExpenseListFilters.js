import React from 'react'
import { DateRangePicker } from 'react-dates'
import { connect } from 'react-redux'
import { setTextFilter, sortBy, availableSortBys, setStartDate, setEndDate } from '../actions/filters'

export class ExpenseListFilters extends React.Component {
    state = {
        calendarFocused: null
    }
    onDatesChange = ({ startDate, endDate }) => {
        this.props.setStartDate(startDate)
        this.props.setEndDate(endDate)
    }
    onFocusChange = (calendarFocused) => {
        this.setState(() => ({ calendarFocused }))
    }
    onSortChange = (e) => {
        this.props.sortBy(e.target.value)
    }
    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value)
    }
    render() {
        return (
            <div>
                <input type="text" value={this.props.filters.text} onChange={this.onTextChange} />
                <select value={this.props.filters.sortBy} onChange={this.onSortChange}>
                    {availableSortBys.map((sortBy) => (
                        <option key={sortBy.value} value={sortBy.value}>{sortBy.description}</option>
                    ))}
                </select>
                <DateRangePicker
                    startDate={this.props.filters.startDate}
                    endDate={this.props.filters.endDate}
                    onDatesChange={this.onDatesChange}
                    focusedInput={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                    showClearDates={true}
                />
            </div>
        )
    }
}

const mapStateToProps = (state) => ({
    filters: state.filters
})

const mapDispatchToProps = (dispatch) => ({
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortBy: (sortBy) => dispatch(sortBy(sortBy)),
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate))
})

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters)