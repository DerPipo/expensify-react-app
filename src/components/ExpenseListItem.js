import React from 'react'
import { connect } from 'react-redux'
import { Link } from 'react-router-dom'
import { removeExpense } from '../actions/expenses'
import formatCurrency from '../utils/currency-format'
import formatDate from '../utils/date-format'

const ExpenseListItem = (props) => (
    <div>
        <Link to={`/edit/${props.expense.id}`}>
            <h3>{props.expense.description}</h3>
        </Link>
        <p>
            {formatCurrency(props.expense.amount)}
            - 
            {formatDate(props.expense.createdAt)}
        </p>
        <button onClick={() => {
            // console.log('Should remove', props.expense.id)
            props.dispatch(removeExpense(props.expense.id))
        }}>Remove</button>
    </div>
)

// export default connect()(ExpenseListItem)
export default ExpenseListItem