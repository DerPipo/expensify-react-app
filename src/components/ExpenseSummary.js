import React from 'react'
import { connect } from 'react-redux'
import selectExpenses from '../selectors/expenses'
import selectExpensesTotal from '../selectors/expenses-total'
import formatCurrency from '../utils/currency-format'

export const ExpenseSummary = (props) => (
    <div>
        <p>Viewing {props.expenseCount} expense{props.expenseCount === 1 ? '' : 's'} totalling {formatCurrency(props.expensesTotal)}</p>
    </div>
)

const mapStateToProps = (state) => {
    const selectedExpenses = selectExpenses(state.expenses, state.filters)
    return {
        expenseCount: selectedExpenses.length,
        expensesTotal: selectExpensesTotal(selectedExpenses)
    }
}

export default connect(mapStateToProps)(ExpenseSummary)