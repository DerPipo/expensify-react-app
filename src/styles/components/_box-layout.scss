.box-layout {
    align-items: center;
    background-image: url('/images/bg.jpg');
    background-size: cover;
    background-position: center;
    display: flex;
    justify-content: center;
    height: 100vh;
}

.box-layout__box {
    background: fade-out(white, 0.15);
    border-radius: 3px;
    padding: $l-size $m-size;
    text-align: center;
    width: 30rem;
}

.box-layout__title {
    line-height: 1;
    margin: 0 0 $m-size 0;
}