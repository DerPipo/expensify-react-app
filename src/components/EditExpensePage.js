import React from 'react';
import { connect } from 'react-redux';
import Modal from 'react-modal';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';
import ExpenseForm from './ExpenseForm';

export class EditExpensePage extends React.Component {
  state = { modalIsOpen: false };
  onSubmit = expense => {
    this.props.startEditExpense(this.props.expense.id, expense);
    this.props.history.push('/');
  };
  onRemoveExpenseButtonClick = () => {
    this.setState({ modalIsOpen: true });
  };
  onConfirmDeleteButtonClick = () => {
    this.props.startRemoveExpense(this.props.expense.id);
    this.props.history.push('/');
  };
  closeModal = () => {
    this.setState({ modalIsOpen: false });
  };
  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">Edit Expense</h1>
          </div>
        </div>
        <div className="content-container">
          <ExpenseForm onSubmit={this.onSubmit} expense={this.props.expense} />
          <button
            className="button button--secondary"
            onClick={this.onRemoveExpenseButtonClick}
          >
            Remove Expense
          </button>
        </div>
        <Modal
          isOpen={this.state.modalIsOpen}
          appElement={document.getElementById('app')}
          onRequestClose={this.closeModal}
          className="modal"
        >
          <h1 className="modal__title">Are you sure?</h1>
          <button
            className="modal__button button"
            onClick={this.onConfirmDeleteButtonClick}
          >
            Yes, delete it
          </button>
          <button
            className="modal__button button button--secondary"
            onClick={this.closeModal}
          >
            Cancel
          </button>
        </Modal>
      </div>
    );
  }
}

const mapStateToProps = (state, props) => ({
  expense: state.expenses.find(expense => expense.id === props.match.params.id),
});

const mapDispatchToProps = dispatch => ({
  startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
  startRemoveExpense: id => dispatch(startRemoveExpense(id)),
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
