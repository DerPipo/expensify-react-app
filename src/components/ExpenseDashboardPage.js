import React from 'react';
import { connect } from 'react-redux';
import AddDummyDataModal from './AddDummyDataModal';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';
import ExpenseSummary from './ExpenseSummary';

export const ExpenseDashboardPage = props => {
  return (
    <div>
      <ExpenseSummary />
      <ExpenseListFilters />
      <ExpenseList />
      {props.noExpenses && <AddDummyDataModal />}
    </div>
  );
};

const mapStateToProps = state => ({
  noExpenses: state.expenses.length === 0,
});

export default connect(mapStateToProps)(ExpenseDashboardPage);
