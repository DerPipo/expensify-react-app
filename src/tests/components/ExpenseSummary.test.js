import React from 'react'
import { shallow } from "enzyme"
import { ExpenseSummary } from "../../components/ExpenseSummary"

test('should render 0 expenses correctly', () => {
    const result = shallow(<ExpenseSummary expenseCount={0} expensesTotal={0} />)
    expect(result).toMatchSnapshot()
})

test('should render 1 expense correctly', () => {
    const result = shallow(<ExpenseSummary expenseCount={1} expensesTotal={1234} />)
    expect(result).toMatchSnapshot()
})

test('should render 2 expenses correctly', () => {
    const result = shallow(<ExpenseSummary expenseCount={2} expensesTotal={100050} />)
    expect(result).toMatchSnapshot()
})
