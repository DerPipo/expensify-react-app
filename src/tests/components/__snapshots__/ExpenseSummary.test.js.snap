// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render 0 expenses correctly 1`] = `
<div>
  <p>
    Viewing 
    0
     expense
    s
     totalling 
    $0.00
  </p>
</div>
`;

exports[`should render 1 expense correctly 1`] = `
<div>
  <p>
    Viewing 
    1
     expense
     totalling 
    $12.34
  </p>
</div>
`;

exports[`should render 2 expenses correctly 1`] = `
<div>
  <p>
    Viewing 
    2
     expense
    s
     totalling 
    $1,000.50
  </p>
</div>
`;
